tests/Makefile

```diff
@@ -0,0 +1,79 @@
+comma := ,
+colon := :
+space :=
+space +=
+
+define colorecho
+	@tput setaf 6
+	@echo $1
+	@tput sgr0
+endef
+
+define capture-output
+	script $1 $2 > /dev/null || cat $1
+endef
+
+DM_JENKINS_REPO ?= ..
+VIRTUALENV_ROOT := $(DM_JENKINS_REPO)/venv
+PATH := $(VIRTUALENV_ROOT)/bin:$(PATH)
+SHELL = bash
+
+SRCDIR := $(DM_JENKINS_REPO)/job_definitions
+TEST_CACHE := .test_cache
+
+ANSIBLE_OUTDIR := $(TEST_CACHE)/job_builder_definitions
+JJB_OUTDIR := $(TEST_CACHE)/jenkins_job_definitions
+
+ifdef JOBS
+	JOBS_LIST := $(subst $(comma),$(space),$(JOBS))
+else
+	JOBS_LIST := $(patsubst %.yml,%,$(notdir $(wildcard $(SRCDIR)/*.yml)))
+endif
+
+SOURCES := $(foreach JOB,$(JOBS_LIST),$(SRCDIR)/$(JOB).yml)
+TARGETS := $(foreach JOB,$(JOBS_LIST),$(JJB_OUTDIR)/$(JOB).xml)
+
+test-job-templates: $(TARGETS)
+
+pre-populate-cache: $(ANSIBLE_OUTDIR)
+
+define process-templates
+	ansible-playbook \
+		--inventory="localhost," \
+		--extra-vars=@$(DM_JENKINS_REPO)/playbooks/roles/jenkins/defaults/main.yml \
+		--extra-vars=@$(DM_JENKINS_REPO)/tests/fixtures/vars.yml \
+		--extra-vars="dest=$1" \
+		--extra-vars="items=[$(subst $(space),$(comma),$2)]" \
+		$(DM_JENKINS_REPO)/tests/test_templates.yml
```

Comment:
I think we should state more clearly the scope of what we're testing. It doesn't pick up bad yaml, for example (changing `concurrent: true` to `concurrent: bananas` was ignored) or any syntax problems within node/shell `script`. That's fine, but we should be clear to the user what a green test implies.

Also, I ran this once and got a bunch of warnings, but re-running gives nothing - is there an option where I can clear that cache?

Reply:
> I think we should state more clearly the scope of what we're testing.

Good point. That might also encourage further tests to be added in future!

> (changing concurrent: true to concurrent: bananas was ignored)

That's a bit concerning, although both are valid YAML I would expect only one of those to be valid for Jenkins.

> Also, I ran this once and got a bunch of warnings, but re-running gives nothing - is there an option where I can clear that cache?

I ran into the same thing, I wasn't sure how to deal with it. It would be nice for the output to be more like a linter, in that it always checks everything, but then it's kind of nice to have the ability to run `make test` and have only the changed files checked.

Reply:
I guess if we wanted it to act more like a linter then the complicated Makefile wouldn't be needed, it could be a more straightforward script... That's an option I hadn't considered before.

Opinions more than welcome.

Reply:
> > (changing concurrent: true to concurrent: bananas was ignored)
> 
> That's a bit concerning, although both are valid YAML I would expect only one of those to be valid for Jenkins.

Might be worth working [jenkins-job-linter](https://github.com/OddBloke/jenkins-job-linter) into the tests as well.

Reply:
👍 for separate linting - ideally as a first step in `make test` (i.e. following our other linting conventions)